### ISSUE TITLE
numpy.log does not return the right result for a complex type masked array

### ISSUE THREAD

**Issue by empet** [numpy/numpy#8516]
H=np.array([[-1.0+0.5j, 2+3j],
                    [0.3-4j, 5+27j]])

    H=ma.masked_where(np.abs(H)>10, H)
    print H
    [[(-1+0.5j) (2+3j)]
     [(0.3-4j) --]]
     print np.log(H)
    [[-- (1.2824746787307684+0.982793723247329j)]
    [(1.3890989805021459-1.4959364790841299j) --]]

**Comment by mhvk** [comment]
The problem is in how the "domain" is defined: it assumes that only values larger than 0 are OK: https://github.com/numpy/numpy/blob/master/numpy/ma/core.py#L1226

Sadly, the problem is more general:
```
In [10]: np.sqrt(-1 + 0j)
Out[10]: 1j

In [11]: np.sqrt(np.ma.MaskedArray(-1 + 0j))
Out[11]: masked
```
It seems like the various domain checkers need to become aware of complex numbers...

**Comment by empet** [comment]
@mhvk   For a real argument, yes the argument must be positive, but for complex numbers such a restriction is a nonsense.

**Comment by mhvk** [comment]
Yes, I agree.  This is definitely a bug! As is insisting on a positive number for `np.sqrt`.
